### ISSUE TITLE
API limit exceeded

### ISSUE THREAD

**Issue by comic96** [joeferraro/MavensMate#597]
Hi,

we have many developers using MavensMate and since the winter 16 upgrade lots of them have reported hitting API limit exceeded problems.

Our DE orgs have a 24 hour limit of 15000 API calls. We have an open case with Salesforce but so far they are not able to offer any advice.

Could there be any issues in the way MM is using the tooling API that have changed recently which might be causing us to hit this limit sooner than we would have done?

Salesforce did report back saying that within a 24 hour period, they could see ~ 7000 request to ="/services/data/v31.0/tooling/query/"

We did an experiment where we saved a vf page to the org, and watched the API usage, even though MM had come back saying the page was saved, we could see the API usage increasing and it didn't stop increasing until we shutdown MM.

Thanks

**Comment by steals** [comment]
Hello @comic96, what version Mavensmate / Mavensmate APP do you use?
Can you setup "SILLY" debug level in the Mavensmate Global Settings ("mm_log_level": "SILLY", in ~/.mavensmate-config.json) and post any relevant logs?

**Comment by comic96** [comment]
Hi @steals the developers in question are using 6.0.0

I got the following information from Salesforce from their splunk logs if that is helpful. In the meantime I will ask the devs to turn on logging.

I just ran the report again for last 24 hrs and mentioned below are my findings :

# API Request Limit Override	default: 15,000 (11,046) - Implied that 11046 is consumed as of now.

# Mentioned below is the detailed description which only includes top 10 values. It consists of 3 columns, first being the entrypoint, second being count and third being %consumption

/services/data/v31.0/tooling/query/	6,893	92.474%	
Api	383	5.138%	
/services/data/v35.0/tooling/query/	83	1.113%	
/services/data/v31.0/tooling/sobjects/ContainerAsyncRequest	40	0.537%	
/services/data/v31.0/tooling/sobjects/ApexPageMember	31	0.416%	
/services/data/v31.0/tooling/sobjects/ApexClassMember	10	0.134%	
/services/data/v31.0/tooling/sobjects/ApexPageMember/402j0000000l9bFAAQ	3	0.04%	
/services/data/v31.0/tooling/sobjects/MetadataContainer	3	0.04%	
/services/data/v35.0/tooling/sobjects/IDEWorkspace/1dej0000001qRejAAE	2	0.027%	
/services/data/v31.0/tooling/sobjects/ApexClassMember/400j0000000nK33AAE	1	0.013%

**Comment by ralphcallaway** [comment]
@comic96 as far as i'm aware there weren't any big changes with v6 that would have altered the api usage

this is kind of a shot in the dark, but personally i've experienced some very slow performance with salesforce compiling things, if that's what you're running into, it's quite possible that mavensmate is just having to poll that much more than usual to get the results of the operation, enough to hit the api limits

that said, all the IDEs are pretty api heavy, and dev orgs are pretty low on api limits, i've had multiple occassions over the years where a small team of developers (e.g. 3) doing a lot of dev work at the same time was able to generate enough api activity to exceed the limit

in both those cases i was able to ask salesforce to increase the api limit temporarily

so in summary, this isn't all that unusual for salesforce development in general, and when it happens they're usually pretty quick to raise the api limits for sandboxes

**Comment by brosiusad** [comment]
I just started experiencing this a couple days ago as well.  I've exceeded the API limit of 15,000 for two days in a row, hampering my ability to get much done.  And as @ralphcallaway points out, I do believe it's due to slow compilation, causing MM to poll for results excessively.  However, if I open mavensmate.log while I'm going a compilation, I can watch **MM poll SF ~12 times a second**.  

Thus, over a particularly slow compilation of, say 2 minutes, I am seeing upwards of 1,440 API calls.  Just to compile one class!  

In one case, I waited 4-5 minutes for a class to compile and ate about a quarter of my API calls for the day.  I also figured out that quitting ST doesn't stop polling; I had to quit MM-app to do that.  Until I figured that out, I had eaten through thousands more API calls.

I think two things are in play: for some reason, compilations are taking longer in SF (especially when there are errors to be reported) and MM polls A LOT.  Is it possible for MM to poll a little less frequently?  Even once per second is more than enough, in my opinion.  That being said, if I peek backwards into mavensmate.log, it seems that polling was happening less frequently before.  Why would it have increased?

Here's a snippet of log from last night (from a compilation I eventually killed by quitting MM-app):

`2016-02-07T03:01:40.060Z - debug: Processing request GET for URL: /status?id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.060Z - debug: query params id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.295Z - debug: Processing request GET for URL: /status?id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.295Z - debug: query params id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.487Z - debug: Processing request GET for URL: /status?id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.487Z - debug: query params id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.569Z - debug: Processing request GET for URL: /status?id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.569Z - debug: query params id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.804Z - debug: Processing request GET for URL: /status?id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.804Z - debug: query params id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.993Z - debug: Processing request GET for URL: /status?id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:40.993Z - debug: query params id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.074Z - debug: Processing request GET for URL: /status?id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.075Z - debug: query params id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.308Z - debug: Processing request GET for URL: /status?id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.308Z - debug: query params id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.502Z - debug: Processing request GET for URL: /status?id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.502Z - debug: query params id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.579Z - debug: Processing request GET for URL: /status?id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.579Z - debug: query params id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.812Z - debug: Processing request GET for URL: /status?id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:41.812Z - debug: query params id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.010Z - debug: Processing request GET for URL: /status?id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.011Z - debug: query params id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.084Z - debug: Processing request GET for URL: /status?id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.084Z - debug: query params id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.319Z - debug: Processing request GET for URL: /status?id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.319Z - debug: query params id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.516Z - debug: Processing request GET for URL: /status?id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.517Z - debug: query params id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.590Z - debug: Processing request GET for URL: /status?id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.590Z - debug: query params id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.823Z - debug: Processing request GET for URL: /status?id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:42.823Z - debug: query params id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.020Z - debug: Processing request GET for URL: /status?id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.020Z - debug: query params id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.098Z - debug: Processing request GET for URL: /status?id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.098Z - debug: query params id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.330Z - debug: Processing request GET for URL: /status?id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.331Z - debug: query params id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.528Z - debug: Processing request GET for URL: /status?id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.528Z - debug: query params id=18d9f450-cd46-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.604Z - debug: Processing request GET for URL: /status?id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.604Z - debug: query params id=d7025540-cd45-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.836Z - debug: Processing request GET for URL: /status?id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
2016-02-07T03:01:43.836Z - debug: query params id=cd4ac340-cd43-11e5-bbff-d541f5e576dc
`

MM app v0.0.10
MM v7.0.0-beta.1
OSX 10.11.3

**Comment by simonlewisBG** [comment]
I think I experienced this issue yesterday. Attempted to save and compile a class back to the server and it was just stuck pending (= bouncing back and forth). Waited a few minutes before closing down Sublime and the MM App. Restarted them both and attempted the Save and Compile again - same as before, it's just sat pending. I left it for about 10 mins before giving up turning off my machine and leaving it for the day. Checked the Org this morning and I've blown through the 15K API usage - normal usage is usually in range of 500-1500 a day.

MM plugin version: 7 - tooling API Enabled
MM App version: v0.0.11-beta1
OSX 10.10.4
